### PR TITLE
Fix underlying type of anonymous classes

### PR DESCRIPTION
**Pull request by smillst**
The underlying type of anonymous classes was the type of it's super class.  For example, the underlying type of `new Object(){}` was Object, but it should be `<OuterClass$1>`.  (daikon and plume-lib type check with this change.)

### CODE REVIEW

File: framework/src/org/checkerframework/framework/type/AnnotatedTypeFactory.java [hunk 1]
```diff
@@ -1934,23 +1934,23 @@ private boolean isSyntheticArgument(Tree tree) {
...
-            // If newClassTree does not create anonymous class,
-            // newClassTree.getIdentifier includes the explicit annotations in this location:
-            //   new @HERE Class()
-            type = (AnnotatedDeclaredType) fromTypeTree(newClassTree.getIdentifier());
-        } else {
-            type = (AnnotatedDeclaredType) toAnnotatedType(InternalUtils.typeOf(newClassTree), false);
-        }
 
         if (newClassTree.getClassBody() != null) {
+            type = (AnnotatedDeclaredType) toAnnotatedType(InternalUtils.typeOf(newClassTree), false);
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/type/AnnotatedTypeFactory.java]: Can an anonymous class use diamonds?
I don't quite understand the three cases and why diamonds are handled again on line 1956.

**Comment by wmdietl** [comment]
Is it possible to write test cases for this?

File: framework/src/org/checkerframework/framework/type/AnnotatedTypeFactory.java (hunk 1 above)
Comment (smillst) [inline framework/src/org/checkerframework/framework/type/AnnotatedTypeFactory.java]: > Can an anonymous class use diamonds?

No
>I don't quite understand the three cases and why diamonds are handled again on line 1956.

I refactored the code to clarify this.

**Comment by smillst** [comment]
@wmdietl I would like to rebase this pull request into one commit before merging.

File: checker/src/org/checkerframework/checker/fenum/FenumVisitor.java [hunk 2]
```diff
@@ -58,7 +59,7 @@ public Void visitSwitch(SwitchTree node, Void p) {
 
     @Override
     protected boolean checkConstructorInvocation(AnnotatedDeclaredType dt,
```
Comment (wmdietl) [inline checker/src/org/checkerframework/checker/fenum/FenumVisitor.java]: Most of the other `checkXXX` methods have a void return type. Is anyone actually using the boolean returned by this function?

**Comment by wmdietl** [comment]
Looks good.

**Comment by smillst** [comment]
merged in ae78785

**Pull request closed without merging**